File: is52038b-labs/16/cpp/Xorshift3520Test.cpp
#include <cppunit/extensions/HelperMacros.h>
#include <cppunit/CompilerOutputter.h>
#include <cppunit/extensions/TestFactoryRegistry.h>
#include <cppunit/ui/text/TestRunner.h>

#include "Xorshift.hpp"

class Xorshift3520Test : public CppUnit::TestFixture {
  CPPUNIT_TEST_SUITE(Xorshift3520Test);
  CPPUNIT_TEST(testSequence);
  CPPUNIT_TEST(testNext);
  CPPUNIT_TEST(testSeed);
  CPPUNIT_TEST_SUITE_END();
public:
  void setUp() { };
  void tearDown() { };

  void testSequence();
  void testNext();
  void testSeed();
};

void Xorshift3520Test::testSequence() {
  Xorshift g = Xorshift(3, 5, 20, 1);
  {
    uint32_t z[10] = {9437193u, 32835u, 1687463497u, 3357546657u, 959044420u, 1628715375u, 2995281287u, 1692197938u, 4175043471u, 595235272u};
    for (int i = 0; i < 10; i++) {
      CPPUNIT_ASSERT_EQUAL_MESSAGE("unexpected next", z[i], g.next());
    }
  }
}

void Xorshift3520Test::testNext() {
  Xorshift g = Xorshift(3, 5, 20, 11);
  uint32_t z = 11;
  for (int i = 0; i < 256; i++) {
    z ^= z << 3;
    z ^= z >> 5;
    z ^= z << 20;
    uint32_t x = g.next();
    CPPUNIT_ASSERT_EQUAL_MESSAGE("unexpected next", z, x);
  }
}

void Xorshift3520Test::testSeed() {
  Xorshift g = Xorshift(3, 5, 20, 11);
  g.seed(1);
  CPPUNIT_ASSERT_EQUAL_MESSAGE("seed appears not to have seeded", 9437193u, g.next());
}

CPPUNIT_TEST_SUITE_NAMED_REGISTRATION(Xorshift3520Test, "3520");
CPPUNIT_REGISTRY_ADD_TO_DEFAULT("3520");

File: is52038b-labs/16/java/Xorshift3520Test.java
import static org.junit.Assert.assertEquals;
import org.junit.Test;

public class Xorshift3520Test {
    @Test
    public void testSequence() {
        Xorshift g = new Xorshift(3, 5, 20, 1);
        {
            long z[] = {9437193L, 32835L, 1687463497L, 3357546657L, 959044420L, 1628715375L, 2995281287L, 1692197938L, 4175043471L, 595235272L};
            for (int i = 0; i < 10; i++) {
                assertEquals("unexpected next", z[i], g.next());
            }
        }
    }
    @Test
    public void testNext() {
        Xorshift g = new Xorshift(3, 5, 20, 11);
        long z = 11;
        for (int i = 0; i < 256; i++) {
	    z ^= (z << 3) % 4294967296L;
	    z ^= (z >> 5) % 4294967296L;
	    z ^= (z << 20) % 4294967296L;
            long x = g.next();
            assertEquals("unexpected next", z, x);
        }
    };
    @Test
    public void testSeed() {
        Xorshift g = new Xorshift(3, 5, 20, 11);
        g.seed(1);
        assertEquals("seed appears not to have seeded", 9437193L, g.next());
    };
}
